File: android/src/main/kotlin/com/gazepoint/flutter/GazepointSdkPlugin.kt
package com.gazepoint.flutter

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.PointF
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import com.gazepoint.sdk.camera.GazeCamera
import com.gazepoint.sdk.camera.GazeCameraOptions
import com.gazepoint.sdk.camera.GazeFrame
import com.gazepoint.sdk.camera.GazePreviewView
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry

/**
 * Flutter wrapper around the Android GazePoint SDK ([GazeCamera]).
 */
class GazepointSdkPlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler,
    ActivityAware,
    PluginRegistry.RequestPermissionsResultListener {

    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var applicationContext: Context? = null
    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null
    private var gazeCamera: GazeCamera? = null
    private var eventSink: EventChannel.EventSink? = null
    private var pendingPermissionResult: Result? = null
    private var pendingPreview: GazePreviewView? = null
    private var isInitialized = false
    private var latestResult: Map<String, Any?>? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "gazepoint_sdk")
        channel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "gazepoint_sdk/gaze_stream")
        eventChannel.setStreamHandler(this)
        binding.platformViewRegistry.registerViewFactory(
            "gazepoint_sdk/preview",
            GazePreviewFactory(this)
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        gazeCamera?.dispose()
        gazeCamera = null
        applicationContext = null
        isInitialized = false
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityBinding?.removeRequestPermissionsResultListener(this)
        activity = null
        activityBinding = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        activityBinding?.removeRequestPermissionsResultListener(this)
        gazeCamera?.stop()
        activity = null
        activityBinding = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    internal fun attachPreview(view: GazePreviewView) {
        pendingPreview = view
        gazeCamera?.attachPreview(view)
    }

    internal fun detachPreview(view: GazePreviewView) {
        if (pendingPreview === view) {
            pendingPreview = null
        }
        gazeCamera?.detachPreview(view)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                val context = applicationContext
                if (context == null) {
                    result.error("NO_CONTEXT", "Plugin not attached", null)
                    return
                }
                val previewEnabled = call.argument<Boolean>("previewEnabled") ?: false
                val showFaceBoxes = call.argument<Boolean>("showFaceBoxes") ?: true
                if (gazeCamera == null) {
                    gazeCamera = GazeCamera(context) { frame ->
                        val mapped = toMap(frame)
                        latestResult = mapped
                        activity?.runOnUiThread {
                            eventSink?.success(mapped)
                        }
                    }
                }
                gazeCamera?.configure(
                    GazeCameraOptions(
                        previewEnabled = previewEnabled,
                        showFaceBoxes = showFaceBoxes
                    )
                )
                pendingPreview?.let { gazeCamera?.attachPreview(it) }
                isInitialized = true
                result.success(null)
            }

            "startTracking" -> {
                val owner = activity as? LifecycleOwner
                if (owner == null) {
                    result.error("NO_ACTIVITY", "Activity not available", null)
                    return
                }
                if (!isInitialized || gazeCamera == null) {
                    result.error("NOT_INITIALIZED", "Call initialize() first", null)
                    return
                }
                gazeCamera?.start(owner)
                result.success(null)
            }

            "stopTracking" -> {
                gazeCamera?.stop()
                result.success(null)
            }

            "setPreviewEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                gazeCamera?.previewEnabled = enabled
                result.success(null)
            }

            "switchCamera" -> {
                gazeCamera?.switchCamera()
                result.success(null)
            }

            "getLatestGaze" -> {
                result.success(latestResult)
            }

            "calibrate" -> {
                val rawPoints = call.argument<List<Map<String, Any>>>("calibrationPoints")
                if (rawPoints == null || rawPoints.size < 3) {
                    result.error("INVALID_ARGS", "At least 3 calibration points required", null)
                    return
                }
                val points = rawPoints.mapNotNull { map ->
                    val expected = map["expected"] as? Map<*, *>
                    val actual = map["actual"] as? Map<*, *>
                    if (expected == null || actual == null) return@mapNotNull null
                    val expectedX = (expected["x"] as? Number)?.toFloat() ?: return@mapNotNull null
                    val expectedY = (expected["y"] as? Number)?.toFloat() ?: return@mapNotNull null
                    val actualX = (actual["x"] as? Number)?.toFloat() ?: return@mapNotNull null
                    val actualY = (actual["y"] as? Number)?.toFloat() ?: return@mapNotNull null
                    PointF(expectedX, expectedY) to PointF(actualX, actualY)
                }
                if (points.size < 3) {
                    result.error("INVALID_ARGS", "Invalid calibration point format", null)
                    return
                }
                gazeCamera?.calibrate(points)
                result.success(null)
            }

            "resetCalibration" -> {
                gazeCamera?.resetCalibration()
                result.success(null)
            }

            "getPerformanceMetrics" -> {
                val metrics = gazeCamera?.getPerformanceMetrics()
                if (metrics == null) {
                    result.error("NOT_INITIALIZED", "Tracker not initialized", null)
                } else {
                    result.success(
                        mapOf(
                            "fps" to metrics.fps.toDouble(),
                            "avgProcessingTimeMs" to metrics.avgProcessingTimeMs.toDouble(),
                            "maxProcessingTimeMs" to metrics.maxProcessingTimeMs.toDouble(),
                            "droppedFrames" to metrics.droppedFrames,
                            "totalFrames" to metrics.totalFrames
                        )
                    )
                }
            }

            "isSupported" -> {
                val context = applicationContext
                if (context == null) {
                    result.success(false)
                    return
                }
                val hasCamera = context.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)
                val hasFront = context.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_FRONT)
                result.success(hasCamera && hasFront)
            }

            "hasCameraPermission" -> {
                val context = applicationContext
                if (context == null) {
                    result.success(false)
                    return
                }
                val granted = ContextCompat.checkSelfPermission(
                    context,
                    Manifest.permission.CAMERA
                ) == PackageManager.PERMISSION_GRANTED
                result.success(granted)
            }

            "requestCameraPermission" -> {
                val currentActivity = activity
                if (currentActivity == null) {
                    result.error("NO_ACTIVITY", "Activity not available", null)
                    return
                }
                if (ContextCompat.checkSelfPermission(
                        currentActivity,
                        Manifest.permission.CAMERA
                    ) == PackageManager.PERMISSION_GRANTED
                ) {
                    result.success(true)
                    return
                }
                pendingPermissionResult = result
                ActivityCompat.requestPermissions(
                    currentActivity,
                    arrayOf(Manifest.permission.CAMERA),
                    REQUEST_CAMERA
                )
            }

            else -> result.notImplemented()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ): Boolean {
        if (requestCode != REQUEST_CAMERA) return false
        val granted = grantResults.isNotEmpty() &&
            grantResults[0] == PackageManager.PERMISSION_GRANTED
        pendingPermissionResult?.success(granted)
        pendingPermissionResult = null
        return true
    }

    private fun toMap(frame: GazeFrame): Map<String, Any?> {
        val gaze = frame.gaze
        val map = mutableMapOf<String, Any?>(
            "faceDetected" to frame.faceDetected,
            "faceCount" to frame.faceCount,
            "statusText" to frame.statusText,
            "timestamp" to System.currentTimeMillis()
        )
        if (gaze != null) {
            map["gazePointX"] = gaze.gazePoint.x.toDouble()
            map["gazePointY"] = gaze.gazePoint.y.toDouble()
            map["confidence"] = gaze.confidence.toDouble()
            map["isBlinking"] = gaze.isBlinking
            map["headPose"] = mapOf(
                "pitch" to gaze.headPose.pitch.toDouble(),
                "yaw" to gaze.headPose.yaw.toDouble(),
                "roll" to gaze.headPose.roll.toDouble()
            )
        }
        return map
    }

    companion object {
        private const val REQUEST_CAMERA = 19283
    }
}

File: macos/gazepoint_sdk/Sources/gazepoint_sdk/GazepointSdkPlugin.swift
import AVFoundation
import Cocoa
import FlutterMacOS

/**
 Flutter plugin wrapper around the macOS GazePoint SDK snapshot.
 Camera preview, face boxes, and tracking live in GazeCamera.
 */
@available(macOS 13.0, *)
public class GazepointSdkPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    private var camera: GazeCamera?
    private var isInitialized = false
    private var latestResult: [String: Any]?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = GazepointSdkPlugin()
        instance.camera = GazeCamera.create()
        instance.camera?.onFrame = { [weak instance] frame in
            let mapped = instance?.toMap(frame)
            instance?.latestResult = mapped
            DispatchQueue.main.async {
                instance?.eventSink?(mapped)
            }
        }

        let methodChannel = FlutterMethodChannel(
            name: "gazepoint_sdk",
            binaryMessenger: registrar.messenger
        )
        instance.methodChannel = methodChannel
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: "gazepoint_sdk/gaze_stream",
            binaryMessenger: registrar.messenger
        )
        instance.eventChannel = eventChannel
        eventChannel.setStreamHandler(instance)

        registrar.register(
            GazePreviewFactory(plugin: instance),
            withId: "gazepoint_sdk/preview"
        )
    }

    func previewView() -> GazePreviewView? {
        camera?.previewView
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            let args = call.arguments as? [String: Any]
            let previewEnabled = args?["previewEnabled"] as? Bool ?? false
            let showFaceBoxes = args?["showFaceBoxes"] as? Bool ?? true
            if camera == nil {
                let cam = GazeCamera.create()
                cam.onFrame = { [weak self] frame in
                    let mapped = self?.toMap(frame)
                    self?.latestResult = mapped
                    DispatchQueue.main.async {
                        self?.eventSink?(mapped)
                    }
                }
                camera = cam
            }
            camera?.options = GazeCameraOptions(
                previewEnabled: previewEnabled,
                showFaceBoxes: showFaceBoxes
            )
            isInitialized = true
            result(nil)

        case "startTracking":
            guard isInitialized, let camera else {
                result(FlutterError(code: "NOT_INITIALIZED", message: "Call initialize() first", details: nil))
                return
            }
            camera.start()
            result(nil)

        case "stopTracking":
            camera?.stop()
            result(nil)

        case "setPreviewEnabled":
            let enabled = (call.arguments as? [String: Any])?["enabled"] as? Bool ?? false
            camera?.previewEnabled = enabled
            result(nil)

        case "switchCamera":
            camera?.switchCamera()
            result(nil)

        case "getLatestGaze":
            result(latestResult)

        case "calibrate":
            guard let args = call.arguments as? [String: Any],
                  let rawPoints = args["calibrationPoints"] as? [[String: Any]],
                  rawPoints.count >= 3
            else {
                result(FlutterError(
                    code: "INVALID_ARGS",
                    message: "At least 3 calibration points required",
                    details: nil
                ))
                return
            }

            var points: [(expected: CGPoint, actual: CGPoint)] = []
            for map in rawPoints {
                guard let expected = map["expected"] as? [String: Any],
                      let actual = map["actual"] as? [String: Any],
                      let ex = expected["x"] as? NSNumber,
                      let ey = expected["y"] as? NSNumber,
                      let ax = actual["x"] as? NSNumber,
                      let ay = actual["y"] as? NSNumber
                else { continue }
                points.append((
                    expected: CGPoint(x: CGFloat(truncating: ex), y: CGFloat(truncating: ey)),
                    actual: CGPoint(x: CGFloat(truncating: ax), y: CGFloat(truncating: ay))
                ))
            }

            guard points.count >= 3 else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid calibration point format", details: nil))
                return
            }
            camera?.calibrate(calibrationPoints: points)
            result(nil)

        case "resetCalibration":
            camera?.resetCalibration()
            result(nil)

        case "getPerformanceMetrics":
            guard let camera else {
                result(FlutterError(code: "NOT_INITIALIZED", message: "Tracker not initialized", details: nil))
                return
            }
            let metrics = camera.getPerformanceMetrics()
            result([
                "fps": Double(metrics.fps),
                "avgProcessingTimeMs": Double(metrics.avgProcessingTimeMs),
                "maxProcessingTimeMs": Double(metrics.maxProcessingTimeMs),
                "droppedFrames": metrics.droppedFrames,
                "totalFrames": metrics.totalFrames
            ])

        case "isSupported":
            result(AVCaptureDevice.default(for: .video) != nil)

        case "hasCameraPermission":
            result(AVCaptureDevice.authorizationStatus(for: .video) == .authorized)

        case "requestCameraPermission":
            let status = AVCaptureDevice.authorizationStatus(for: .video)
            switch status {
            case .authorized:
                result(true)
            case .notDetermined:
                AVCaptureDevice.requestAccess(for: .video) { granted in
                    DispatchQueue.main.async { result(granted) }
                }
            default:
                result(false)
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    private func toMap(_ frame: GazeFrame) -> [String: Any] {
        var map: [String: Any] = [
            "faceDetected": frame.faceDetected,
            "faceCount": frame.faceCount,
            "statusText": frame.statusText,
            "timestamp": Int(Date().timeIntervalSince1970 * 1000)
        ]
        if let gaze = frame.gaze {
            map["gazePointX"] = Double(gaze.gazePoint.x)
            map["gazePointY"] = Double(gaze.gazePoint.y)
            map["confidence"] = Double(gaze.confidence)
            map["isBlinking"] = gaze.isBlinking
            map["headPose"] = [
                "pitch": Double(gaze.headPose.pitch),
                "yaw": Double(gaze.headPose.yaw),
                "roll": Double(gaze.headPose.roll)
            ]
        }
        return map
    }
}

@available(macOS 13.0, *)
final class GazePreviewFactory: NSObject, FlutterPlatformViewFactory {
    private weak var plugin: GazepointSdkPlugin?

    init(plugin: GazepointSdkPlugin) {
        self.plugin = plugin
        super.init()
    }

    func createArgsCodec() -> (FlutterMessageCodec & NSObjectProtocol)? {
        FlutterStandardMessageCodec.sharedInstance()
    }

    func create(withViewIdentifier viewId: Int64, arguments args: Any?) -> NSView {
        let container = NSView(frame: .zero)
        container.wantsLayer = true
        container.layer?.backgroundColor = NSColor.black.cgColor
        if let preview = plugin?.previewView() {
            preview.removeFromSuperview()
            preview.translatesAutoresizingMaskIntoConstraints = true
            preview.autoresizingMask = [.width, .height]
            preview.frame = container.bounds
            container.addSubview(preview)
        }
        return container
    }
}
